File: matlab/Nifty.m
classdef Nifty
    properties
        q
        int
        err
        dmax
        Nr = 50
        isZeroAtDmax = true
        isZeroAtZero = true
    end
    properties(Dependent = true)
        F
        r
        dr
        k
        L
        A
        Nq
        b
    end
    methods
        function obj = Nifty()
        end
        function val = get.F(obj)
            val = 4*pi*obj.dr*sinc(obj.q(:)*obj.r'/pi);
            val(:,[1,end]) = 0.5*val(:,[1,end]); % trapezoid rule
        end
        function val = get.k(obj)
            val = obj.Nr - obj.isZeroAtDmax - obj.isZeroAtZero;
        end
        function val = get.dr(obj)
            val = obj.dmax/(obj.Nr-1);
        end
        function val = get.r(obj)
            val = obj.dmax*linspace(0,1,obj.Nr)';
        end
        function val = get.Nq(obj)
            val = length(obj.q);
        end
        function val = get.L(obj)
            val = sparse(1:(obj.Nr-2),1:(obj.Nr-2),-.5,obj.Nr-2,obj.Nr) + ...
                sparse(1:(obj.Nr-2),2:(obj.Nr-1),1,obj.Nr-2,obj.Nr) + ...
                sparse(1:(obj.Nr-2),3:(obj.Nr),-.5,obj.Nr-2,obj.Nr);
            if obj.isZeroAtDmax
                val = val(:,1:(end-1));
            end
            if obj.isZeroAtZero
                val = val(:,2:end);
            end
        end
        function val = get.A(obj)
            val = obj.F;
            if length(obj.err)==1
                val = val/obj.err;
            elseif isempty(obj.err)
                % do nothing
            else
                val = repmat(1./obj.err(:),1,obj.Nr).*val;
            end
            if obj.isZeroAtDmax
                val = val(:,1:(end-1));
            end
            if obj.isZeroAtZero
                val = val(:,2:end);
            end
        end
        function val = get.b(obj)
            val = obj.int(:)./obj.err(:);
        end
        function [w,alpha] = ift(obj,alpha)
            Lmat = obj.L;
            Amat = obj.A;
            AA = Amat'*Amat;
            LL = full(Lmat'*Lmat);
            if isempty(alpha)
                alpha = trace(AA)/trace(LL);
            end
            w = (AA + alpha*LL)\(Amat'*obj.b);
            if obj.isZeroAtDmax
                w = [w;0];
            end
            if obj.isZeroAtZero
                w = [0;w];
            end
        end

        function [y,y0,w0] = make_test_data(obj,sigma)
            if nargin==1 || isempty(sigma)
                sigma = obj.err;
            end
            x = obj.r/obj.dmax;
            w0 = 1-(2*x-1).^2;
            w0 = w0.*x.*exp(-2*x);
            w0 = w0 / (4*pi*obj.dr*sum(w0)); % normalize (I(0) = 1)
            y0 = obj.F*w0;
            if ~isempty(sigma)
                y = y0  + randn([obj.Nq,1]).*sigma;
            else
                y = [];
            end
        end
    end
        
    methods(Static)
        function [r,w,ireg,alpha] = sprite(q,int,err,varargin)
                         
            N = Nifty();
            N.q = q;
            N.int = int;
            N.err = err;
            
            % assign optional arguments
            for j=1:2:length(varargin)
                N.(varargin{j}) = varargin{j+1};
            end
            
            [w,alpha] = N.ift([]);
            r = N.r;
            ireg = N.F*w;
        end
    end
end

File: python/nifty.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from scipy import sparse as sp

class nifty:
    
    def __init__(self):
        self.q = None
        self.intensity = None
        self.err = None
        self.dmax = None
        self.Nr = 50
        self.is_zero_at_dmax = True
        self.is_zero_at_zero = True
        
    @property
    def F(self):
        F_tmp = 4 * np.pi * self.dr * np.sinc(np.outer(self.q,self.r) / np.pi)
        F_tmp[:,[0,-1]] = 0.5*F_tmp[:,[0,-1]]
        return F_tmp
    
    @property
    def k(self):
        return self.Nr - int(self.is_zero_at_dmax) - int(self.is_zero_at_zero)
    
    @property
    def dr(self):
        return self.dmax / (self.Nr - 1)
    
    @property
    def r(self):
        return self.dmax * np.linspace(0,1,self.Nr)
    
    @property
    def Nq(self):
        return len(self.q)
    
    @property
    def L(self):
        L_tmp = sp.csr_matrix((-0.5*np.ones(self.Nr-2), (np.arange(0,self.Nr-2), np.arange(0,self.Nr-2))), shape=(self.Nr-2, self.Nr))+\
            sp.csr_matrix((np.ones(self.Nr-2), (np.arange(0,self.Nr-2), np.arange(1,self.Nr-1))), shape=(self.Nr-2, self.Nr))+\
            sp.csr_matrix((-0.5*np.ones(self.Nr-2), (np.arange(0,self.Nr-2), np.arange(2,self.Nr))), shape=(self.Nr-2, self.Nr))
        if self.is_zero_at_dmax:
            L_tmp = L_tmp[:,:-1]
        if self.is_zero_at_zero:
            L_tmp = L_tmp[:,1:]
        return L_tmp
    
    @property
    def A(self): #Did not consider the case when err is a constant
        A_tmp = self.F
        if len(self.err) != 0:
            A_tmp = (1 / self.err[:,np.newaxis] * np.ones(self.Nr)) * A_tmp
        if self.is_zero_at_dmax:
            A_tmp = A_tmp[:,:-1]
        if self.is_zero_at_zero:
            A_tmp = A_tmp[:,1:]
        return A_tmp
    
    @property
    def b(self):
        return self.intensity/self.err
    
    def ift(self, alpha=None):
        Lmat = self.L
        Amat = self.A
        AA = Amat.T @ Amat
        LL = sp.csr_matrix.todense(Lmat.T @ Lmat)
        if alpha == None:
            alpha = np.trace(AA)/np.trace(LL)
        w = np.linalg.lstsq(AA + alpha * LL, Amat.T @ self.b, rcond=None)[0]
        if self.is_zero_at_dmax:
            w = np.append(w, 0)
        if self.is_zero_at_zero:
            w = np.insert(w, 0, 0)
        return [w, alpha]
    
    @staticmethod
    def sprite(q,intensity,err,dmax,Nr = 50,is_zero_at_dmax = True,is_zero_at_zero = True):
        N = nifty()
        N.q = q
        N.intensity = intensity
        N.err = err
        N.dmax = dmax
        N.Nr = Nr
        N.is_zero_at_dmax = is_zero_at_dmax
        N.is_zero_at_zero = is_zero_at_zero
        
        [w, alpha] = N.ift()
        r = N.r
        ireg = N.F @ w
        
        return [r, w, ireg, alpha]
        
